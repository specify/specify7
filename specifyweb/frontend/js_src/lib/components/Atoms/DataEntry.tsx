import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import { commonText } from '../../localization/common';
import { formsText } from '../../localization/forms';
import type { RA } from '../../utils/types';
import type { AnySchema } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import type { ViewDescription } from '../FormParse';
import type { cellAlign, cellVerticalAlign } from '../FormParse/cells';
import { Button } from './Button';
import { className } from './className';
import type { icons } from './Icons';
import { Link } from './Link';
import type { TagProps } from './wrapper';
import { wrap } from './wrapper';

const dataEntryButton = (
  className: string,
  title: LocalizedString,
  icon: keyof typeof icons
) =>
  function (
    props: Omit<TagProps<'button'>, 'children' | 'type'> & {
      readonly onClick:
        | ((event: React.MouseEvent<HTMLButtonElement>) => void)
        | undefined;
    }
  ): JSX.Element {
    return (
      <Button.Icon
        className={`${className} ${props.className ?? ''}`}
        icon={icon}
        title={title}
        {...props}
      />
    );
  };

export const columnDefinitionsToCss = (
  columns: RA<number | undefined>,
  flexibleColumnWidth: boolean
): string =>
  columns
    .map((width) =>
      typeof width === 'number'
        ? `${width}${flexibleColumnWidth ? 'fr' : 'px'}`
        : 'auto'
    )
    .join(' ');

/**
 * Components for Specify Form
 * This is called DataEntry instead of Form because "Form" is already taken
 */

export const DataEntry = {
  Grid: wrap<
    'div',
    {
      readonly viewDefinition: ViewDescription;
      readonly flexibleColumnWidth: boolean;
      readonly display: 'block' | 'inline';
    }
  >(
    'DataEntry.Grid',
    'div',
    `items-center p-1 -ml-1 gap-2`,
    ({
      viewDefinition,
      display,
      className: classNameString = '',
      flexibleColumnWidth,
      style,
      ...props
    }) => ({
      className: `${
        display === 'inline' ? 'inline-grid' : 'grid'
      } ${classNameString}`,
      style: {
        gridTemplateColumns: columnDefinitionsToCss(
          viewDefinition.columns,
          flexibleColumnWidth
        ),
        ...style,
      },
      ...props,
    })
  ),
  Header: wrap('DataEntry.Header', 'header', className.formHeader),
  Title: wrap(
    'DataEntry.Title',
    'h2',
    `${className.headerPrimary} ${className.formTitle}`
  ),
  Cell: wrap<
    'div',
    {
      readonly colSpan: number;
      readonly align: typeof cellAlign[number];
      readonly visible: boolean;
      readonly verticalAlign: typeof cellVerticalAlign[number];
    }
  >(
    'DataEntry.Cell',
    'div',
    'flex flex-col',
    ({ colSpan, align, visible, verticalAlign, ...props }) => ({
      ...props,
      style: {
        visibility: visible ? undefined : 'hidden',
        gridColumn:
          colSpan === 1 ? undefined : `span ${colSpan} / span ${colSpan}`,
        alignItems:
          align === 'right'
            ? 'flex-end'
            : align === 'center'
            ? 'center'
            : undefined,
        alignSelf:
          verticalAlign === 'stretch'
            ? 'stretch'
            : verticalAlign === 'center'
            ? 'self-center'
            : verticalAlign === 'start'
            ? 'self-start'
            : verticalAlign === 'end'
            ? 'self-end'
            : undefined,
        ...props.style,
      },
    })
  ),
  Footer: wrap('DataEntry.Footer', 'div', className.formFooter, {
    role: 'toolbar',
  }),
  SubForm: wrap('DataEntry.SubForm', 'fieldset', 'contents'),
  SubFormHeader: wrap(
    'DataEntry.SubFormHeader',
    'legend',
    'gap-2 flex font-bold border-b border-gray-500 pt-5 pb-1 items-center',
    ({ children, ...props }) => ({
      // A hack for Safari. See https://github.com/specify/specify7/issues/1535
      children: <div {...props}>{children}</div>,
    })
  ),
  SubFormTitle: wrap('DataEntry.SubFormTitle', 'h3', className.formTitle),
  Add: dataEntryButton(className.dataEntryAdd, commonText.add(), 'plus'),
  View: dataEntryButton(className.dataEntryView, commonText.view(), 'eye'),
  Edit: dataEntryButton(className.dataEntryEdit, commonText.edit(), 'pencil'),
  Clone: dataEntryButton(
    className.dataEntryClone,
    formsText.clone(),
    'clipboard'
  ),
  Search: dataEntryButton(
    className.dataEntrySearch,
    commonText.search(),
    'search'
  ),
  Remove: dataEntryButton(
    className.dataEntryRemove,
    commonText.remove(),
    'minus'
  ),
  Visit({
    className: localClassName = '',
    resource,
  }: {
    readonly className?: string;
    readonly resource: SpecifyResource<AnySchema> | undefined;
  }): JSX.Element | null {
    return typeof resource === 'object' && !resource.isNew() ? (
      <Link.NewTab
        aria-label={commonText.openInNewTab()}
        className={`${className.dataEntryVisit} ${localClassName}`}
        href={resource.viewUrl()}
        title={commonText.openInNewTab()}
      />
    ) : null;
  },
};
/* eslint-enable @typescript-eslint/naming-convention */
