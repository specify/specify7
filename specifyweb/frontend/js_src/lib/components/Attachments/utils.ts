import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
