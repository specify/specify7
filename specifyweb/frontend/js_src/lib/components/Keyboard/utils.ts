import type { LocalizedString } from 'typesafe-i18n';

import type { RA } from '../../utils/types';
import { localized } from '../../utils/types';
import type { KeyboardShortcuts, ModifierKey } from './config';
import { keyboardModifierLocalization, keyboardPlatform } from './config';
import { keyJoinSymbol } from './context';

const localizedKeyJoinSymbol = ' + ';
export const localizeKeyboardShortcut = (shortcut: string): LocalizedString =>
  localized(
    shortcut
      .split(keyJoinSymbol)
      .map((key) => keyboardModifierLocalization[key as ModifierKey] ?? key)
      .join(localizedKeyJoinSymbol)
  );

/**
 * If there is a keyboard shortcut defined for current system, use it
 * (also, if current system explicitly has empty array of shortcuts, use it).
 *
 * Otherwise, use the keyboard shortcut from one of the other platforms if set,
 * but change meta to ctrl and vice versa as necessary.
 */
export function resolvePlatformShortcuts(
  shortcut: KeyboardShortcuts
): RA<string> | undefined {
  if ('platform' in shortcut) return shortcut[keyboardPlatform];
  else if ('other' in shortcut)
    return keyboardPlatform === 'windows'
      ? shortcut.other
      : shortcut.other?.map(replaceCtrlWithMeta);
  else if ('windows' in shortcut)
    return keyboardPlatform === 'other'
      ? shortcut.other
      : shortcut.other?.map(replaceCtrlWithMeta);
  else if ('mac' in shortcut) return shortcut.other?.map(replaceMetaWithCtrl);
  else return undefined;
}

const replaceCtrlWithMeta = (shortcut: string): string =>
  shortcut
    .split(keyJoinSymbol)
    .map((key) => (key === 'ctrl' ? 'meta' : key))
    .join(keyJoinSymbol);

const replaceMetaWithCtrl = (shortcut: string): string =>
  shortcut
    .split(keyJoinSymbol)
    .map((key) => (key === 'meta' ? 'ctrl' : key))
    .join(keyJoinSymbol);
