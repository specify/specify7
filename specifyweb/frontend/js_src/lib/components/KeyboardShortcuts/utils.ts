import type { LocalizedString } from 'typesafe-i18n';

import type { RA, WritableArray } from '../../utils/types';
import { localized } from '../../utils/types';
import type { KeyboardShortcuts, ModifierKey } from './config';
import {
  keyboardModifierLocalization,
  keyboardPlatform,
  keyLocalizations,
  shiftKeyLocalizations,
} from './config';
import { keyJoinSymbol } from './context';

export const localizedKeyJoinSymbol = ' + ';
export function localizeKeyboardShortcut(shortcut: string): LocalizedString {
  const parts = shortcut.split(keyJoinSymbol);
  const hasShift = parts.includes('Shift');

  const modifiers: WritableArray<string> = [];
  const nonModifiers: WritableArray<string> = [];
  // eslint-disable-next-line functional/no-loop-statement
  for (const key of parts) {
    const localizedModifier = keyboardModifierLocalization[key as ModifierKey];
    if (typeof localizedModifier === 'string')
      modifiers.push(localizedModifier);
    else {
      nonModifiers.push(
        (hasShift ? shiftKeyLocalizations[key] : undefined) ??
          keyLocalizations[key] ??
          key
      );
    }
  }

  // If there is only one non-modifier key, then join the keys without separator
  const resolved =
    nonModifiers.length > 1
      ? [...modifiers, ...nonModifiers].join(localizedKeyJoinSymbol)
      : `${modifiers.join('')}${nonModifiers[0]}`;

  return localized(resolved);
}

/**
 * If there is a keyboard shortcut defined for current system, use it
 * (also, if current system explicitly has empty array of shortcuts, use it).
 *
 * Otherwise, use the keyboard shortcut from one of the other platforms if set,
 * but change meta to ctrl and vice versa as necessary.
 */
export function resolvePlatformShortcuts(
  shortcut: KeyboardShortcuts
): RA<string> | undefined {
  if (keyboardPlatform in shortcut) return shortcut[keyboardPlatform];
  else if ('other' in shortcut)
    return keyboardPlatform === 'windows'
      ? shortcut.other
      : shortcut.other?.map(replaceCtrlWithMeta);
  else if ('windows' in shortcut)
    return keyboardPlatform === 'other'
      ? shortcut.other
      : shortcut.other?.map(replaceCtrlWithMeta);
  else if ('mac' in shortcut) return shortcut.other?.map(replaceMetaWithCtrl);
  else return undefined;
}

const replaceCtrlWithMeta = (shortcut: string): string =>
  shortcut
    .split(keyJoinSymbol)
    .map((key) => (key === 'Ctrl' ? 'Meta' : key))
    .join(keyJoinSymbol);

const replaceMetaWithCtrl = (shortcut: string): string =>
  shortcut
    .split(keyJoinSymbol)
    .map((key) => (key === 'Meta' ? 'Ctrl' : key))
    .join(keyJoinSymbol);
