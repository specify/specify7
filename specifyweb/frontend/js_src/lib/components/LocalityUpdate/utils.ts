import type { LocalizedString } from 'typesafe-i18n';

import { commonText } from '../../localization/common';
import { localityText } from '../../localization/locality';
import { f } from '../../utils/functools';
import type { IR, RA, RR } from '../../utils/types';
import { tables } from '../DataModel/tables';
import type { Tables } from '../DataModel/types';
import { resolveBackendParsingMessage } from '../WorkBench/resultsParser';
import type { LocalityUpdateHeader, LocalityUpdateTaskStatus } from './types';

const localityUpdateAcceptedLocalityFields: RA<
  Lowercase<keyof Tables['Locality']['fields']>
> = ['guid', 'datum', 'latitude1', 'longitude1'];

export const localityUpdateAcceptedHeaders = f.store(
  () =>
    new Set([
      ...localityUpdateAcceptedLocalityFields,
      ...tables.GeoCoordDetail.literalFields
        .map(({ name }) => name.toLowerCase())
        .filter((header) => header !== 'locality'),
    ])
);

export const localityUpdateRequiredHeaders = new Set<LocalityUpdateHeader>([
  'guid',
]);

export const localityUpdateStatusLocalization: RR<
  LocalityUpdateTaskStatus,
  LocalizedString
> = {
  PARSED: localityText.localityUpdateParsed(),
  PARSING: localityText.localityUpdateParsing(),
  PENDING: localityText.localityUpdateStarting(),
  PROGRESS: localityText.localityUpdateProgressing(),
  FAILED: localityText.localityUpdateFailed(),
  PARSE_FAILED: localityText.localityUpdateParseFailure(),
  ABORTED: localityText.localityUpdateCancelled(),
  SUCCEEDED: localityText.localityUpdateSucceeded(),
};

export function resolveLocalityUpdateErrorMessage(
  key: string,
  payload: IR<unknown>
): LocalizedString {
  const baseParseResults = resolveBackendParsingMessage(key, payload);

  if (baseParseResults !== undefined) {
    return baseParseResults;
  } else if (key === 'guidHeaderNotProvided') {
    return localityText.guidHeaderNotProvided();
  } else if (key === 'noLocalityMatchingGuid') {
    return localityText.noLocalityMatchingGuid({
      guid: payload.guid as string,
    });
  } else if (key === 'multipleLocalitiesWithGuid') {
    return localityText.multipleLocalitiesWithGuid({
      guid: payload.guid as string,
      localityIds: (payload.localityIds as RA<number>).join(', '),
    });
  } else {
    return commonText.colonLine({
      label: key,
      value: Object.keys(payload).length === 0 ? '' : JSON.stringify(payload),
    });
  }
}
