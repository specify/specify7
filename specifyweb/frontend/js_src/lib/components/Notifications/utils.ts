import { ping } from '../../utils/ajax/ping';
import { GenericNotification } from './NotificationRenderers';
import { formData } from '../../utils/ajax/helpers';
import { f } from '../../utils/functools';

export function deleteNotification(
  notification: GenericNotification
): Promise<void> {
  return ping('/notifications/delete/', {
    method: 'POST',
    // eslint-disable-next-line @typescript-eslint/naming-convention
    body: formData({ message_id: notification.messageId }),
    errorMode: 'dismissible',
  }).then(f.void);
}
