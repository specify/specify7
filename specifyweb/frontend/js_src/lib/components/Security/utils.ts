import type { LocalizedString } from 'typesafe-i18n';

import { userText } from '../../localization/user';
import { ajax } from '../../utils/ajax';
import { Http } from '../../utils/ajax/definitions';
import { f } from '../../utils/functools';
import type { IR, RA } from '../../utils/types';
import {
  camelToHuman,
  lowerToHuman,
  replaceKey,
  sortFunction,
  toLowerCase,
} from '../../utils/utils';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { strictGetTable, tables } from '../DataModel/tables';
import type { Tables } from '../DataModel/types';
import {
  frontEndPermissions,
  operationPolicies,
  tableActions,
} from '../Permissions/definitions';
import type { RoleBase } from './Collection';
import { processPolicies } from './policyConverter';
import { getRegistriesFromPath } from './registry';
import type { Role } from './Role';

export type BackEndRole = Omit<Role, 'policies'> & {
  readonly policies: IR<RA<string>>;
};

export const fetchRoles = async (
  collectionId: number
): Promise<RA<Role> | undefined> =>
  ajax<RA<BackEndRole>>(`/permissions/roles/${collectionId}/`, {
    headers: { Accept: 'application/json' },
    /*
     * When looking at a different collection, it is not yet know if user has
     * read permission. Instead of waiting for permission query to complete,
     * query anyway and silently handle the permission denied error
     */
    expectedErrors: [Http.FORBIDDEN],
  }).then(({ data, status }) =>
    status === Http.FORBIDDEN
      ? undefined
      : data
          .map((role) => ({
            ...role,
            policies: processPolicies(role.policies),
          }))
          .sort(sortFunction(({ name }) => name))
  );

export const fetchUserRoles = async (
  collectionId: number,
  userId: number
): Promise<RA<RoleBase> | undefined> =>
  ajax<RA<{ readonly id: number; readonly name: LocalizedString }>>(
    `/permissions/user_roles/${collectionId}/${userId}`,
    {
      headers: { Accept: 'application/json' },
      /*
       * When looking at a different collection, it is not yet know if user has
       * read permission. Instead of waiting for permission query to complete,
       * query anyway and silently handle the permission denied error
       */
      expectedErrors: [Http.FORBIDDEN],
    }
  ).then(({ data, status }) =>
    status === Http.FORBIDDEN
      ? undefined
      : data
          .map(({ id, name }) => ({ roleId: id, roleName: name }))
          /*
           * Sort all roles by ID, so that can easier detect if user roles changed
           * Since last save
           */
          .sort(sortFunction(({ roleId }) => roleId))
  );

/**
 * Convert a path like "/table/agent" into "Table > Agent"
 */
export function resourceNameToLongLabel(resource: string): string {
  const parts = resourceNameToParts(resource);
  return parts
    .map((_, index) =>
      resourceNameToLabel(partsToResourceName(parts.slice(0, index + 1)))
    )
    .join(' > ');
}

/**
 * Convert "/table" into "Table" and "/table/agent" into "Agent"
 */
export function resourceNameToLabel(resource: string): LocalizedString {
  if (
    /*
     * "getRegistriesFromPath" does not work for system tables that are part
     * of a tool, so have to handle that case here
     */
    resource.startsWith(tablePermissionsPrefix) &&
    !resource.includes(anyResource)
  )
    return resourceNameToTable(resource).label;
  else {
    const parts = resourceNameToParts(resource);
    return (
      getRegistriesFromPath(parts)[parts.length - 1]?.[parts.at(-1)!]?.label ??
      camelToHuman(resource)
    );
  }
}

/** Like getRegistriesFromPath, but excludes institutional policies */
export function getCollectionRegistriesFromPath(resourceParts: RA<string>) {
  const registries = getRegistriesFromPath(resourceParts);
  return registries.map((part, index) =>
    part === undefined
      ? undefined
      : Object.fromEntries(
          Object.entries(part).map(([resource, data]) => [
            resource,
            /*
             * Put institutional policy into a separate group, unless
             * it is selected
             */
            data.isInstitutional && resource !== resourceParts[index]
              ? replaceKey(
                  data,
                  'groupName',
                  userText.excludedInstitutionalPolicies()
                )
              : data,
          ])
        )
  );
}

/**
 * Localize action name
 */
export const actionToLabel = (action: string): LocalizedString =>
  action === anyAction ? userText.allActions() : lowerToHuman(action);

export const toolPermissionPrefix = 'tools';
export const anyAction = '%';
export const anyResource = '%';
export const permissionSeparator = '/';

export const resourceNameToParts = (resourceName: string): RA<string> =>
  resourceName.split(permissionSeparator).filter(Boolean);

export const resourceNameToTable = (resourceName: string): SpecifyTable =>
  strictGetTable(resourceNameToParts(resourceName)[1]);

export const partsToResourceName = (parts: RA<string>): string =>
  parts.length === 1 && parts[0] === anyResource
    ? anyResource
    : `${permissionSeparator}${parts.join(permissionSeparator)}`;

export const tablePermissionsPrefix = `${permissionSeparator}table${permissionSeparator}`;
export const tableNameToResourceName = <TABLE_NAME extends keyof Tables>(
  tableName: TABLE_NAME
): `${typeof tablePermissionsPrefix}${Lowercase<TABLE_NAME>}` =>
  `${tablePermissionsPrefix}${toLowerCase(tableName)}`;

/**
 * Special resource that is needed by the back-end for user to be able to
 * edit anything.
 * Field level permissions are not yet fully implemented, thus this resource
 * must be hidden in the UI, but present in all policy lists
 */
export const fieldPolicy = '/field/%';

/**
 * Front-end enforces that each user that has collection access, also has the
 * following permissions:
 */
export const basicPermissions: IR<RA<string>> = {
  [fieldPolicy]: [anyAction],
};

/**
 * Get a union of all actions that can be done on descendants of a given
 * permission resource type
 */
export function getAllActions(rawPath: string): RA<string> {
  if (rawPath.startsWith(`${permissionSeparator}${toolPermissionPrefix}`))
    return tableActions;
  else {
    const parts = resourceNameToParts(rawPath);
    const path = partsToResourceName(
      parts.at(-1) === anyResource ? parts.slice(0, -1) : parts
    );
    return f.unique(
      [
        ...Object.entries(operationPolicies),
        ...Object.entries(frontEndPermissions),
        ...Object.keys(tables).map(
          (tableName) =>
            [tableNameToResourceName(tableName), tableActions] as const
        ),
      ]
        .filter(([key]) => key.startsWith(path))
        .flatMap(([_key, actions]) => actions)
    );
  }
}
