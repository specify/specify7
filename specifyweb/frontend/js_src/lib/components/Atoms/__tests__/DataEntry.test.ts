import { snapshot } from '../../../tests/reactUtils';
import { theories } from '../../../tests/utils';
import { f } from '../../../utils/functools';
import type { AnySchema } from '../../DataModel/helperTypes';
import type { SpecifyResource } from '../../DataModel/legacyTypes';
import type { SpecifyTable } from '../../DataModel/specifyTable';
import { columnDefinitionsToCss, DataEntry } from '../DataEntry';

theories(columnDefinitionsToCss, [
  { in: [[1, 2, 3], true], out: '1fr 2fr 3fr' },
  { in: [[1, 2, undefined, 3], false], out: '1px 2px auto 3px' },
]);

snapshot(DataEntry.Grid, {
  viewDefinition: {
    rows: [],
    columns: [1, 2, 3, undefined],
    formType: 'form',
    mode: 'edit',
    table: undefined as unknown as SpecifyTable,
    name: 'test',
  },
  flexibleColumnWidth: true,
  display: 'block',
  children: 'Test',
});

snapshot(DataEntry.Header, { children: 'Test' });
snapshot(DataEntry.Title, { children: 'Test' });

describe('DataEntry.Cell', () => {
  snapshot(
    DataEntry.Cell,
    {
      children: 'Test',
      colSpan: 1,
      align: 'right',
      verticalAlign: 'stretch',
      visible: true,
    },
    'colspan 1, align right, self stretch'
  );
  snapshot(
    DataEntry.Cell,
    {
      children: 'Test',
      colSpan: 3,
      align: 'left',
      verticalAlign: 'center',
      visible: true,
    },
    'colspan 3, align left, self center'
  );
  snapshot(
    DataEntry.Cell,
    {
      children: 'Test',
      colSpan: 1,
      align: 'left',
      verticalAlign: 'end',
      visible: false,
    },
    'invisible, self end'
  );
});

snapshot(DataEntry.Footer, { children: 'Test' });
snapshot(DataEntry.SubForm, { children: 'Test' });
snapshot(DataEntry.SubFormHeader, { children: 'Test' });
snapshot(DataEntry.SubFormTitle, { children: 'Test' });
snapshot(DataEntry.Add, { onClick: f.never });
snapshot(DataEntry.View, { onClick: f.never });
snapshot(DataEntry.Edit, { onClick: f.never });
snapshot(DataEntry.Clone, { onClick: f.never });
snapshot(DataEntry.Search, { onClick: f.never });
snapshot(DataEntry.Remove, { onClick: f.never });

describe('DataEntry.visit', () => {
  snapshot(DataEntry.Visit, { resource: undefined }, 'no resource');
  snapshot(
    DataEntry.Visit,
    {
      resource: {
        isNew: () => false,
        viewUrl: () => 'RESOURCE_VIEW_URL',
      } as unknown as SpecifyResource<AnySchema>,
    },
    'no resource'
  );
});
