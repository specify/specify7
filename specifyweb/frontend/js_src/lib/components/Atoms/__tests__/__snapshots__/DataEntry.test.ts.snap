// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` renders without errors 1`] = `
<DocumentFragment>
  <button
    aria-label="Add"
    class="icon link rounded !text-green-700 print:hidden "
    title="Add"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="w-6 h-6 flex-shrink-0"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M10 5a1 1 0 011 1v3h3a1 1 0 110 2h-3v3a1 1 0 11-2 0v-3H6a1 1 0 110-2h3V6a1 1 0 011-1z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[` renders without errors 2`] = `
<DocumentFragment>
  <button
    aria-label="View"
    class="icon link rounded !text-cyan-400 print:hidden "
    title="View"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="w-6 h-6 flex-shrink-0"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10 12a2 2 0 100-4 2 2 0 000 4z"
      />
      <path
        clip-rule="evenodd"
        d="M.458 10C1.732 5.943 5.522 3 10 3s8.268 2.943 9.542 7c-1.274 4.057-5.064 7-9.542 7S1.732 14.057.458 10zM14 10a4 4 0 11-8 0 4 4 0 018 0z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[` renders without errors 3`] = `
<DocumentFragment>
  <button
    aria-label="Edit"
    class="icon link rounded !text-orange-400 print:hidden "
    title="Edit"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="w-6 h-6 flex-shrink-0"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M13.586 3.586a2 2 0 112.828 2.828l-.793.793-2.828-2.828.793-.793zM11.379 5.793L3 14.172V17h2.828l8.38-8.379-2.83-2.828z"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[` renders without errors 4`] = `
<DocumentFragment>
  <button
    aria-label="Clone"
    class="icon link rounded !text-amber-700 print:hidden "
    title="Clone"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="w-6 h-6 flex-shrink-0"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 3a1 1 0 011-1h2a1 1 0 110 2H9a1 1 0 01-1-1z"
      />
      <path
        d="M6 3a2 2 0 00-2 2v11a2 2 0 002 2h8a2 2 0 002-2V5a2 2 0 00-2-2 3 3 0 01-3 3H9a3 3 0 01-3-3z"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[` renders without errors 5`] = `
<DocumentFragment>
  <button
    aria-label="Search"
    class="icon link rounded !text-blue-500 print:hidden "
    title="Search"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="w-6 h-6 flex-shrink-0"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M8 4a4 4 0 100 8 4 4 0 000-8zM2 8a6 6 0 1110.89 3.476l4.817 4.817a1 1 0 01-1.414 1.414l-4.816-4.816A6 6 0 012 8z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[` renders without errors 6`] = `
<DocumentFragment>
  <button
    aria-label="Remove"
    class="icon link rounded !text-red-700 print:hidden "
    title="Remove"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="w-6 h-6 flex-shrink-0"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M3 10a1 1 0 011-1h12a1 1 0 110 2H4a1 1 0 01-1-1z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`DataEntry.Cell colspan 1, align right, self stretch 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
    style="align-items: flex-end; align-self: stretch;"
  >
    Test
  </div>
</DocumentFragment>
`;

exports[`DataEntry.Cell colspan 3, align left, self center 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
    style="grid-column: span 3 / span 3; align-self: self-center;"
  >
    Test
  </div>
</DocumentFragment>
`;

exports[`DataEntry.Cell invisible, self end 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
    style="visibility: hidden; align-self: self-end;"
  >
    Test
  </div>
</DocumentFragment>
`;

exports[`DataEntry.Footer renders without errors 1`] = `
<DocumentFragment>
  <div
    class="border-brand-300 border-t-2 flex print:hidden pt-2 gap-2 flex-wrap justify-center flex-col md:flex-row"
    role="toolbar"
  >
    Test
  </div>
</DocumentFragment>
`;

exports[`DataEntry.Grid renders without errors 1`] = `
<DocumentFragment>
  <div
    class="grid items-center p-1 -ml-1 gap-2"
    style="grid-template-columns: 1fr 2fr 3fr auto;"
  >
    Test
  </div>
</DocumentFragment>
`;

exports[`DataEntry.Header renders without errors 1`] = `
<DocumentFragment>
  <header
    class="border-b-2 border-brand-300 flex items-center pb-2 gap-2 md:gap-4 md:justify-between flex-col md:flex-row"
  >
    Test
  </header>
</DocumentFragment>
`;

exports[`DataEntry.SubForm renders without errors 1`] = `
<DocumentFragment>
  <fieldset
    class="contents"
  >
    Test
  </fieldset>
</DocumentFragment>
`;

exports[`DataEntry.SubFormHeader renders without errors 1`] = `
<DocumentFragment>
  <legend>
    <div
      class="gap-2 flex font-bold border-b border-gray-500 pt-5 pb-1 items-center"
    >
      Test
    </div>
  </legend>
</DocumentFragment>
`;

exports[`DataEntry.SubFormTitle renders without errors 1`] = `
<DocumentFragment>
  <h3
    class="text-lg font-bold flex items-center gap-2"
  >
    Test
  </h3>
</DocumentFragment>
`;

exports[`DataEntry.Title renders without errors 1`] = `
<DocumentFragment>
  <h2
    class="font-semibold text-black dark:text-white text-lg font-bold flex items-center gap-2"
  >
    Test
  </h2>
</DocumentFragment>
`;

exports[`DataEntry.visit no resource 1`] = `<DocumentFragment />`;

exports[`DataEntry.visit no resource 2`] = `
<DocumentFragment>
  <a
    aria-label="Open in New Tab"
    class="link !text-blue-700 print:hidden "
    href="RESOURCE_VIEW_URL"
    rel="noopener"
    target="_blank"
    title="Open in New Tab"
  >
    <span
      title="(opens in a new tab)"
    >
      <span
        class="sr-only"
      >
        (opens in a new tab)
      </span>
      <svg
        aria-hidden="true"
        class="w-6 h-6 flex-shrink-0"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 3a1 1 0 100 2h2.586l-6.293 6.293a1 1 0 101.414 1.414L15 6.414V9a1 1 0 102 0V4a1 1 0 00-1-1h-5z"
        />
        <path
          d="M5 5a2 2 0 00-2 2v8a2 2 0 002 2h8a2 2 0 002-2v-3a1 1 0 10-2 0v3H5V7h3a1 1 0 000-2H5z"
        />
      </svg>
    </span>
  </a>
</DocumentFragment>
`;
